### ISSUE TITLE
GPG_0 triggered after one touch, even if another GPG instance is waiting for a second touch

### ISSUE THREAD

**Issue by n-st** [maximbaz/yubikey-touch-detector#14]
When the user starts multiple GPG instances, each touch only confirms one of them, in the order they were started.
The touch detector, however, already switches back to GPG_0 ("no touch required") after the first GPG instance completes its wait.

To reproduce:
1. Start first instance: `echo | gpg --sign > /dev/null`
2. yubikey-touch-detector: `[notifiers/stdout] GPG_1`
3. Start second instance: `echo | gpg --sign > /dev/null`
4. *touch Yubikey*
5. First instance terminates
6. yubikey-touch-detector: `[notifiers/stdout] GPG_0`
7. *touch Yubikey*
8. Second instance terminates

**Comment by n-st** [comment]
The GPG detector switches back to GPG_0 as soon as `--card-status` has succeeded once:

https://github.com/maximbaz/yubikey-touch-detector/blob/6ff3ff0225d23d8922c1bf74c1fdf6804e20db02/detector/gpg.go#L68-L74

It seems that requests to the card/Yubikey are queued in order, which in the example would be (1) `gpg --sign`, (2) `yubikey-touch-detector`/`gpg --card-status`, (3) `gpg --sign`.

In that case, I think we should be able to detect the waiting (3) `gpg --sign` by launching another `gpg --card-status` right after the first one succeeds:

				checkGPGCardStatus().Wait()
				checkGPGCardStatus().Wait()

Incidentally, that seems to work even for longer queues (three or four simultaneous `gpg --sign`s), though I'm not entirely sure why…

**Comment by maximbaz** [comment]
Thanks for the report, interesting catch 👍 

Just so you know, contrary to U2F / Webauthn touch detection where we have a proper touch detection based on the published specifications, gpg detection is unfortunately based on a bunch of hacky hacks, so certain weird cases are unavoidable (particularly, I'm very disappointed about #8). I'm hoping to meet a brave soul one day who uses this app and knows their way around wireshark to help me implement the proper support for smartcard protocol... 🙂 

Having said that, let's do all we can in the meantime to make our current logic good enough!

Your example is interesting because it doesn't reproduce with encryption, only with signing operation 🤕 

And I confirm that just adding a second `checkGPGCardStatus().Wait()` works even for longer queues, it seems to solve the issue entirely. I cannot explain it either, but it's a great addition to our list of hacks 😜 

I'll commit your proposal 😄

**Comment by n-st** [comment]
Sorry to burst the bubble… I thought about this a bit more and found an example where the quickfix doesn't work:

1. Start first instance: `echo | gpg --sign > /dev/null`
2. yubikey-touch-detector: `[notifiers/stdout] GPG_1`
3. Start second instance: `echo | gpg --sign > /dev/null`
4. _touch Yubikey_
5. First instance terminates
6. **Start third instance: `echo | gpg --sign > /dev/null`**
7. _touch Yubikey_
8. Second instance terminates
9. **yubikey-touch-detector: `[notifiers/stdout] GPG_0`**
10. _touch Yubikey_
11. Third instance terminates

Not surprising, actually: All card access attempts queue in order.
With my earlier example, we queue effectively queue all signing attempts before our second card-status check — and so, quite coincidentally, there are no further ops waiting behind the card-status check.
If there are, the result from the second check is as incorrect as it was before the quickfix.

Looks like we need a loop after all… :confused:

**Comment by maximbaz** [comment]
Nice catch 👍 Do you feel like sending another PR? 😉

**Comment by n-st** [comment]
I'm actually not sure how exactly to solve this, since we don't want to run an infinite loop…

Let see if I've got this right:
https://github.com/maximbaz/yubikey-touch-detector/blob/6ff3ff0225d23d8922c1bf74c1fdf6804e20db02/detector/gpg.go#L68
[start:] If we have been `Wait()`ing for a significant time (longer than 50 or 100 ms maybe*), then the Yubikey was indeed waiting for a touch (if not, the access to pubring.kbx wasn't related to a sign/encrypt op).
In that case, right when we finish `Wait()`ing, we want to check if there were any ops waiting behind our card-status request, i.e. a new card-status request takes longer than those 50 or 100 ms.
If it doesn't take that long, we're done, GPG_0.
If it does take that long, repeat at [start].

… and I'm not sure how to apply that here because I don't know the workings of `<-requestGPGCheck`, `time.AfterFunc`, the cases in which `cmd.Wait()` returns an `err`, and the thought behind the (max.) 20 loop iterations … :sweat_smile: 

<sub>(*`gpg --card-status` takes 0.020 seconds with my Yubikey 5)</sub>

**Comment by maximbaz** [comment]
> [start:] If we have been Wait()ing for a significant time (longer than 50 or 100 ms maybe*), then the Yubikey was indeed waiting for a touch (if not, the access to pubring.kbx wasn't related to a sign/encrypt op).
> (*gpg --card-status takes 0.020 seconds with my Yubikey 5)

Correct, but not only that, the current value is bigger than what you expected because it is hand-picked so that for people like me who disabled touch for a signature operation, yubikey-touch-detector doesn't trigger `GPG_1` while it's signing - and that takes longer than `0.020` seconds, but below `1` second.

Everything else is basically correct 🙂 

What if we instead of doing this (as in master currently):

```
checkGPGCardStatus().Wait()
checkGPGCardStatus().Wait()
```

We would repeat in a loop this:

```
cmd := checkGPGCardStatus()
timer := time.AfterFunc(1000*time.Millisecond, func() {
	cmd.Process.Kill()
})
err := cmd.Wait()
timer.Stop()
```

Until `err` becomes `nil`. When `err` becomes `nil`, it would mean that `gpg --card-status` completed faster than 1s, so no more gpg operation is currently running, and we are happy to declare `GPG_0`.

Makes sense? Can you find a flaw in such logic? 😜

**Comment by n-st** [comment]
> We would repeat in a loop this: […]

That looks correct.
I have a sneaking suspicion that the manually selected wait time is setting us up for a race condition sometime somewhere, but it seems like the best we can do at the moment…

**Comment by maximbaz** [comment]
Would you like to try the current `master`? The logic is this, card is blocked while it's waiting for a touch, but touching yubikey unblocks the card (and if there are more pending gpg operations, the next time you check the card it will again be blocked).

So to really make sure that there are no more pending gpg operations, we wait until we can confirm that card is not blocked twice in a row.

**Comment by n-st** [comment]
Tested with a chain of four instances (sign, sign, touch, sign, touch, sign, touch, touch) and it reports GPG_1 … GPG_0 correctly.
Looks like that's solved, then. :)

Would you like to prepare a new release so it will trickle down to my distro's packager?

**Comment by maximbaz** [comment]
Will do now, thanks for testing 👍 If you are on Arch, then I'm the distro packager, it's a one-man show 😄
